From e0a5284620e8d020771bf8c1fa73d6113e662ebf Mon Sep 17 00:00:00 2001
From: Jim Easterbrook <jim@jim-easterbrook.me.uk>
Date: Wed, 22 Oct 2025 09:04:45 +0100
Subject: [PATCH] Skip more refcount tests if Python >= 3.14

Python 3.14 introduced optimisation of refcounts so they aren't
incremented if Python knows the object is not going to be deleted.
Disabling these tests is temporary until I have a Python 3.14
installation on which to develop better tests.
---
 tests/test_image.py   | 12 ++++++++----
 tests/test_preview.py |  6 ++++--
 tests/test_types.py   |  6 ++++--
 tests/test_value.py   | 12 ++++++++----
 4 files changed, 24 insertions(+), 12 deletions(-)

diff --git a/tests/test_image.py b/tests/test_image.py
index 691a400..7b2dbd6 100644
--- a/tests/test_image.py
+++ b/tests/test_image.py
@@ -121,17 +121,21 @@ def test_Image(self):
         self.check_result(image.iccProfileDefined(), bool, False)
         # test data access
         image.readMetadata()
-        self.assertEqual(sys.getrefcount(image), 2)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(image), 2)
         view = image.data()
-        self.assertEqual(sys.getrefcount(image), 3)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(image), 3)
         self.check_result(view, memoryview, self.image_data)
         self.assertEqual(view.readonly, True)
         image.writeMetadata()
-        self.assertEqual(sys.getrefcount(image), 2)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(image), 2)
         with self.assertRaises(ValueError):
             view[0]
         del view
-        self.assertEqual(sys.getrefcount(image), 2)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(image), 2)
         # test other methods
         image.readMetadata()
         self.check_result(image.byteOrder(),
diff --git a/tests/test_preview.py b/tests/test_preview.py
index 476abbf..f05f2b7 100644
--- a/tests/test_preview.py
+++ b/tests/test_preview.py
@@ -54,9 +54,11 @@ def test_PreviewImage(self):
         with preview.data() as data:
             self.check_result(data, memoryview, copy.data())
             self.assertEqual(data[:10], b'\xff\xd8\xff\xe0\x00\x10JFIF')
-        self.assertEqual(sys.getrefcount(preview), 3)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(preview), 3)
         del data
-        self.assertEqual(sys.getrefcount(preview), 2)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(preview), 2)
         with self.assertWarns(DeprecationWarning):
             self.assertEqual(memoryview(preview), copy.data())
         self.check_result(preview.extension(), str, '.jpg')
diff --git a/tests/test_types.py b/tests/test_types.py
index 5db5143..cd365dc 100644
--- a/tests/test_types.py
+++ b/tests/test_types.py
@@ -84,9 +84,11 @@ def test_DataBuf(self):
         with self.assertRaises(ValueError):
             view[0]
         view = buf.data()
-        self.assertEqual(sys.getrefcount(buf), 3)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(buf), 3)
         del view
-        self.assertEqual(sys.getrefcount(buf), 2)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(buf), 2)
 
     def test_Rational(self):
         for type_ in (exiv2.Rational, exiv2.URational):
diff --git a/tests/test_value.py b/tests/test_value.py
index 8661f7d..d891ac4 100644
--- a/tests/test_value.py
+++ b/tests/test_value.py
@@ -131,9 +131,11 @@ def do_common_string_tests(self, value, data):
         view = copy.data()
         self.assertIsInstance(view, memoryview)
         self.assertEqual(view[0], data[0])
-        self.assertEqual(sys.getrefcount(copy), 3)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(copy), 3)
         del view
-        self.assertEqual(sys.getrefcount(copy), 2)
+        if sys.version_info < (3, 14):
+            self.assertEqual(sys.getrefcount(copy), 2)
 
     def do_common_xmp_tests(self, value):
         with self.assertWarns(DeprecationWarning):
@@ -355,9 +357,11 @@ def check_data(value, data):
                 self.assertEqual(view, data)
             with self.assertRaises(ValueError):
                 self.assertEqual(view[0], data[0])
-            self.assertEqual(sys.getrefcount(value), 3)
+            if sys.version_info < (3, 14):
+                self.assertEqual(sys.getrefcount(value), 3)
             del view
-            self.assertEqual(sys.getrefcount(value), 2)
+            if sys.version_info < (3, 14):
+                self.assertEqual(sys.getrefcount(value), 2)
         else:
             copy = value.data()
             self.assertIsInstance(copy, bytearray)
